server-common/src/config_ext.rs

```diff
@@ -160,7 +160,8 @@ where
         let peer_ip = stream.peer_addr().ok().flatten().map(|addr| addr.ip());
 
         let stream = match self.acceptor.accept(stream).await {
-            Ok(stream) => stream,
+            Ok(Some(stream)) => stream,
+            Ok(None) => return,
             Err(e) => {
```

Comment:
Am I correct that this breaking change is to avoid the log::error line on L165?

Reply:
Yes. Having an `Option` here allows a success that doesn't produce a stream.

I can think of other *possible* ways to make this work, if you specifically don't want to make this breaking change. For example, we could add a defaulted trait method `ignore_err` that takes an error and returns a bool, with the defaulted version always returning false. Would you prefer that?

Reply:
I just wanted to make sure I understood the motivation behind the PR so we can think through possible solutions. I see at least four possible solutions to this, and want to make sure they're each considered. The hesitation is not about making a breaking change, but giving breaking changes sufficient consideration. I usually consider (or implement) multiple solutions and then select the one that seems best. I welcome participation in this process, but generally prefer to start with the constraints and intent, not a solution.

My understanding of the problem: "Some Acceptor implementations only want to log::error when the Input type represents a Transport, and handle other Inputs internally"

Possible solutions, in no particular order:

1. return `Result<Option<Self::Output>>` — this PR
    - advantages: clearly distinguishes between the failure condition and the noop condition. If the error case included further behavior than just logging at some future point, it would be possible to exclude these `Ok(None)` lines
    - disadvantages: unclear for a non-acme tls implementation would need to return an option, possibly a sharp edge when implementing — "should this be a None or an Err" probably needs documentation
2. return `Option<Self::Output>, move the log lines into the acceptor implementation
    - advantages: implementation implications are clear, can remove the Error type for a reduction in complexity. implementing code can make better choices about how and what to log than a generic `Debug + Send + Sync`. Additionally, logging coming from a tls-specific crate makes it slightly easier to adjust the log verbosity with something like env_logger.
    - disadvantages: logging is not handled centrally, and what an implementer might expect to be a Result is a None. This would likely also require some documentation, like "if you have an error, log it internally to the Acceptor implementation"
3. the above-suggested defaulted trait method (`Acceptor::ignore_err(Self::Error)`)
    - advantages: no breaking change
    - disadvantages: awkward complexity just to avoid a log line
4. moving this behavior to the acme Transport, treating this as an `Ok` but returning an acme-specific Transport variant that has a `Poll::Ready(0)` implementation
    - advantages: no breaking changes, no complexity introduced to other acceptor implementations
    - disadvantages: would need to confirm that this does not result in a different log::Error message (will report back)


I'm currently prioritizing them 2≥4>1>3, but welcome further arguments for or against any of them. If 4 is silent, that seems like the least invasive change

Reply:
Updated the above: 4 seems like a viable solution that requires no changes to Acceptor and contains the complexity to the acme crate's implementation. If the Transport type is an enum that has a variant with AsyncRead and AsyncWrite implementations that return `Poll::Ready(Ok(0))`/`Poll::Ready(Ok(()))`, the only highest-log-level output is a
`[2023-10-09T17:58:01Z DEBUG trillium_server_common::config_ext] closing connection` which seems accurate. I'd happily PR this to the acme implementation; is there a public repo?

Reply:
@jbr I haven't written the ACME implementation yet, because I was waiting on the outcome of this PR. :)

I could return an enum for a transport, and have a variant that represents a non-connection. However, that means a generic caller cannot tell that the connection is one that'll never produce input and ignore any output, so as you noted, it'll expect data and not get any, which may result in an error like that. For HTTP, that's *probably* fine, since the caller expects to read data before writing any. That would cause larger problems if ever using this connection interface for any *other* purpose in which the server speaks first, but since this is HTTP-specific, it should work.

I'll give that a try and see how it goes.

Reply:
Also, I think I'd rule out option 2, because doing logging in more places will make it more difficult to change the logging strategy (e.g. using tracing).